File: src/main/java/com/zzc/leetcode_sep/Rob.java
package com.zzc.leetcode_sep;

/**
 * @author zc.zhou
 * @Description
 * @create 2023-09-16 1:06
 */
public class Rob {
    public static void main(String[] args) {

    }

    public int rob(int[] nums) {
        if (nums.length == 1) {
            return nums[0];
        }
        if (nums.length == 2) {
            return Math.max(nums[0], nums[1]);
        }
        int[] money = new int[nums.length];
        money[0] = nums[0];
        money[1] = Math.max(nums[0], nums[1]);
        for (int i = 2; i < money.length; i++) {
            money[i] = Math.max(money[i - 1], nums[i] + money[i - 2]);
        }

        return money[money.length - 1];
    }
}

File: pyproject/dp/Rob.py
from typing import List


class Solution:
    def rob(self, nums: List[int]) -> int:
        size = len(nums)
        dp = [0] * size
        if size == 1:
            return nums[0]
        dp[0] = nums[0]
        dp[1] = max(dp[0], nums[1])
        for i in range(2, size):
            # 考虑第i天是否偷
            # 偷：nums[i] + dp[i-2]; 不偷：dp[i-1]
            dp[i] = max(dp[i - 2] + nums[i], dp[i - 1])
        return dp[size - 1]

s = Solution()
n=[1,2,3,1]
print(s.rob(n))